rasa_nlu/featurizers/ngram_featurizer.py

```diff
@@ -0,0 +1,271 @@
+            clf.fit(X, y)
+            scores = clf.scores_
+            sort_idx = [i[0] for i in sorted(enumerate(scores), key=lambda x: -1 * x[1])]
+
+            return np.array(list_of_ngrams)[sort_idx]
+        else:
+            return []
+
+    def _ngrams_in_sentences(self, sentences, spacy_nlp, ngrams):
+        """Given a set of sentences, returns a vector of {1,0} values indicating ngram presence"""
+
+        all_vectors = []
+        for sentence in sentences:
+            presence_vector = self._ngrams_in_sentence(sentence, spacy_nlp, ngrams)
+            all_vectors.append(presence_vector)
+        return all_vectors
+
+    def _ngrams_in_sentence(self, sentence, spacy_nlp, ngrams):
+        """Given a set of sentences, returns a vector of {1,0} values indicating ngram presence"""
+
+        import numpy as np
+
+        cleaned_sentence = self._remove_in_vocab_words_of_sentence(sentence, spacy_nlp)
+        presence_vector = np.zeros(len(ngrams))
+        idx_array = [idx for idx in xrange(len(ngrams)) if ngrams[idx] in cleaned_sentence]
+        presence_vector[idx_array] = 1
+        return presence_vector
+
+    def _generate_all_ngrams(self, list_of_strings):
+        """Takes a list of strings and generates all character ngrams.
+
+        Generated ngrams are at least 3 characters (and at most 17), occur at least 5 times
+        and occur independently of longer superset ngrams at least once."""
+
+        features = {}
+        counters = {}
+
+        for n in xrange(self.n_gram_min_length, self.n_gram_max_length):
+            candidates = []
+            features[n] = []
+            counters[n] = Counter()
+
+            # generate all possible n length ngrams
+            for text in list_of_strings:
+                text = text.replace(',', ' ')
+                for word in text.lower().split(' '):
+                    cands = [word[i:i + n] for i in xrange(len(word) - n)]
+                    for cand in cands:
+                        counters[n][cand] += 1
+                        if cand not in candidates:
+                            candidates.append(cand)
+
+            # iterate over these candidates picking only the applicable ones
+            for can in candidates:
+                if counters[n][can] > self.n_gram_min_occurrences:
+                    features[n].append(can)
+                    begin = can[:-1]
+                    end = can[1:]
+                    if n > self.n_gram_min_length:
+                        if counters[n - 1][begin] == counters[n][can] and begin in features[n - 1]:
+                            features[n - 1].remove(begin)
+                        if counters[n - 1][end] == counters[n][can] and end in features[n - 1]:
+                            features[n - 1].remove(end)
+
+        return [item for sublist in features.values() for item in sublist]
+
+    def _create_bow_vecs(self, intent_features, sentences, spacy_nlp, max_ngrams=None):
```

Comment:
I agree this should change, however I guess the BoW vecs are already features and we are adding on top of them, so maybe `augment_feature_vecs`?